library/core/src/ptr/mod.rs

```diff
@@ -1135,27 +1135,56 @@ pub const unsafe fn replace<T>(dst: *mut T, mut src: T) -> T {
 #[rustc_const_unstable(feature = "const_ptr_read", issue = "80377")]
 #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
 pub const unsafe fn read<T>(src: *const T) -> T {
-    // We are calling the intrinsics directly to avoid function calls in the generated code
-    // as `intrinsics::copy_nonoverlapping` is a wrapper function.
-    extern "rust-intrinsic" {
-        #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.63.0")]
-        fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
-    }
+    // It would be semantically correct to implement this via `copy_nonoverlapping`
+    // and `MaybeUninit`, as was done before PR #109035.
 
-    let mut tmp = MaybeUninit::<T>::uninit();
-    // SAFETY: the caller must guarantee that `src` is valid for reads.
-    // `src` cannot overlap `tmp` because `tmp` was just allocated on
-    // the stack as a separate allocated object.
+    // However, it switched to intrinsic that lowers to `_0 = *src` in MIR in
+    // order to address a few implementation issues:
     //
-    // Also, since we just wrote a valid value into `tmp`, it is guaranteed
-    // to be properly initialized.
+    // - Using `MaybeUninit::assume_init` after a `copy_nonoverlapping` was not
+    //   turning the untyped copy into a typed load. As such, the generated
+    //   `load` in LLVM didn't get various metadata, such as `!range` (#73258),
+    //   `!nonnull`, and `!noundef`, resulting in poorer optimization.
```

Comment:
```suggestion
    // - Using `MaybeUninit::assume_init` after a `copy_nonoverlapping`, the compiler
    //    was unable to exploit the typed load in `assume_init`. As such, the generated
    //   `load` in LLVM didn't get various metadata, such as `!range` (#73258),
    //   `!nonnull`, and `!noundef`, resulting in poorer optimization.
```
I think it should be clear that this is a compiler issue, not a lang/spec issue. There *is* a typed load in `assume_init`.

Reply:
"unable to exploit" is horribly vague, it can mean anything. It can say "the compiler" but it should be specific about what transform was failing.

Reply:
I'll add some more text up near line 1138 -- that's where I was trying to say that the other approach gives enough information to be allowed to do it, as opposed to this one about implementation issues.

Reply:
Yeah, the key point is that both approaches give enough information, we just fail to make use of it the other way. The `copy_nonoverlapping` is kind of a red herring.

Reply:
Updated to explicitly say the `assume_init` _ought_ to be enough, but "the compiler was not capable of taking advantage of that information".